cpp/src/datetime/datetime_ops.cu

```diff
@@ -41,22 +41,14 @@
 #include <thrust/iterator/permutation_iterator.h>
 #include <thrust/transform.h>
 
+#define extract(field)                                                          \
+  case field:                                                                   \
+    return detail::apply_datetime_op<detail::extract_component_operator<field>, \
+                                     cudf::type_id::INT16>(column, stream, mr)
```

Comment:
When producing the detail function, can you move this definition to closer to the function definition, you can even do it in the function body:

```
std::unique_ptr<column> extract_datetime_component(column_view const& column, datetime_component component, ...)
{
  #define extract(field) \
     case field: return ...;
  switch(component) {
     extract(datetime_component::DAY);
     extract(...);
     default:
       CUDF_FAIL(...);
  }
#undef extract
}